machxo2/bitstream.cc

```diff
@@ -18,264 +18,741 @@
-    else {
-        if (found->second.is_string)
-            return found->second.as_string();
-        else
-            return std::to_string(found->second.as_int64());
+    // Get the list of tiles corresponding to a PLL
+    std::vector<std::string> get_pll_tiles(BelId bel)
+    {
+        std::string name = ctx->tile_info(bel)->bel_data[bel.index].name.get();
+        std::vector<std::string> tiles;
+        Loc loc = ctx->getBelLocation(bel);
+
+        if (name == "LPLL") {
+            tiles.push_back(ctx->get_tile_by_type_loc(loc.y - 1, loc.x - 1, "GPLL_L0"));
+        } else if (name == "RPLL") {
+            tiles.push_back(ctx->get_tile_by_type_loc(loc.y + 1, loc.x - 1, "GPLL_R0"));
+        } else {
+            NPNR_ASSERT_FALSE_STR("bad PLL loc " + name);
+        }
+        return tiles;
     }
-};
 
-// Get the PIC tile corresponding to a PIO bel
-static std::string get_pic_tile(Context *ctx, BelId bel)
-{
-    static const std::set<std::string> pio_t = {"PIC_T0", "PIC_T0_256", "PIC_TS0"};
-    static const std::set<std::string> pio_b = {"PIC_B0", "PIC_B0_256", "PIC_BS0_256"};
-    static const std::set<std::string> pio_l = {"PIC_L0",       "PIC_L1",       "PIC_L2",       "PIC_L3",
-                                                "PIC_LS0",      "PIC_L0_VREF3", "PIC_L0_VREF4", "PIC_L0_VREF5",
-                                                "PIC_L1_VREF3", "PIC_L1_VREF4", "PIC_L1_VREF5", "PIC_L2_VREF4",
-                                                "PIC_L2_VREF5", "PIC_L3_VREF4", "PIC_L3_VREF5"};
-    static const std::set<std::string> pio_r = {"PIC_R0",     "PIC_R1",     "PIC_RS0",
-                                                "PIC_R0_256", "PIC_R1_640", "PIC_RS0_256"};
-
-    std::string pio_name = ctx->tile_info(bel)->bel_data[bel.index].name.get();
-    if (bel.location.y == 0) {
-        return ctx->get_tile_by_type_loc(0, bel.location.x, pio_t);
-    } else if (bel.location.y == ctx->chip_info->height - 1) {
-        return ctx->get_tile_by_type_loc(bel.location.y, bel.location.x, pio_b);
-    } else if (bel.location.x == 0) {
-        return ctx->get_tile_by_type_loc(bel.location.y, 0, pio_l);
-    } else if (bel.location.x == ctx->chip_info->width - 1) {
-        return ctx->get_tile_by_type_loc(bel.location.y, bel.location.x, pio_r);
-    } else {
-        NPNR_ASSERT_FALSE("bad PIO location");
+    void set_pip(ChipConfig &cc, PipId pip)
+    {
+        std::string tile = ctx->get_pip_tilename(pip);
+        std::string tile_type =
+                ctx->chip_info->tiletype_names[ctx->tile_info(pip)->pip_data[pip.index].tile_type].get();
+        std::string source = get_trellis_wirename(pip.location, ctx->getPipSrcWire(pip));
+        std::string sink = get_trellis_wirename(pip.location, ctx->getPipDstWire(pip));
+        cc.tiles[tile].add_arc(sink, source);
+
+        // Special case pips whose config bits are spread across tiles.
+        if (source == "G_PCLKCIBVIQT0" && sink == "G_VPRXCLKI0") {
```

Comment:
could there well be more cases of this?

Reply:
This is actually from old / existing code 
( https://github.com/YosysHQ/nextpnr/blob/master/machxo2/bitstream.cc#L132 ) so far it looks like it is this one that requires copy